### ISSUE TITLE
How to implement model check pointing for best models

### ISSUE THREAD

**Issue by kaustubhharapanahalli** [facebookresearch/detectron2#645]
Hi,

How can I checkpoint a model based on accuracy instead of saving checkpoint based on the number of iterations. Currently, I am using `cfg.SOLVER.CHECKPOINT_PERIOD` which uses number of iterations. What do I have to do if I need to save only the best weights instead of saving model after every 5000 iterations?

**Comment by ppwwyyxx** [comment]
With plain_train_net.py, you can add this logic using the return value of `do_test`: https://github.com/facebookresearch/detectron2/blob/5e2a6f62ef752c8b8c700d2e58405e4bede3ddbe/tools/plain_train_net.py#L175-L182

**Comment by kaustubhharapanahalli** [comment]
Okay, thank you! I'll try with it

**Comment by Kbkartik** [comment]
@ppwwyyxx, how do we implement model checkpointing based on validation loss? While evaluating, method 'do_test' in plain_train_net.py returns AP scores and not the loss.

**Comment by ppwwyyxx** [comment]
The model does not and could not compute a loss when in testing mode. See #499.

**Comment by Kbkartik** [comment]
@ppwwyyxx, by going through similar issues #499 and #569, I see you mentioning the use of a custom model. Could you elaborate on this? Also, is there any other way where I can get the loss on validation set in a object detection scenario?

**Comment by ppwwyyxx** [comment]
The model computes losses when in training mode. So `model(validation_inputs)` will give you losses.

By writing the model yourself you'll be able to decide what it returns if you'd like anything not supported by the existing models.

See also https://detectron2.readthedocs.io/tutorials/models.html#model-output-format and https://detectron2.readthedocs.io/tutorials/write-models.html
